src/pyspiffe/bundle/x509_bundle/x509_bundle.py

```diff
@@ -1,9 +1,261 @@
 """
-X509Bundle module manages X509Bundle objects.
+This module manages X.509 Bundle objects.
 """
 
+import os
+from typing import Set
+
+import pem
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization
+from cryptography.x509 import Certificate
+from typing.io import BinaryIO
+
+from pyspiffe.bundle.x509_bundle.exceptions import (
+    ParseX509BundleError,
+    X509BundleError,
+    LoadX509BundleError,
+    SaveX509BundleError,
+)
+from pyspiffe.spiffe_id.trust_domain import TrustDomain, EMPTY_DOMAIN_ERROR
+from pyasn1.codec.der.decoder import decode
+
+_BUNDLE_FILE_MODE = 0o644
+
+__all__ = ['X509Bundle']
+
 
 class X509Bundle(object):
-    """X509Bundle is a collection of trusted public key material for a trust domain."""
+    """
+    Represents a collection of trusted X.509 authorities for a trust domain.
+    """
+
+    def __init__(
+        self,
+        trust_domain: TrustDomain,
+        x509_authorities: Set[Certificate],
+    ) -> None:
+        """Creates a X509Bundle instance.
+
+        Args:
+            trust_domain: A TrustDomain instance.
+            x509_authorities: A set of CA certificates.
+        """
+        self._trust_domain = trust_domain
+        self._x509_authorities = x509_authorities
+        if not self._x509_authorities:
+            self._x509_authorities = set()
+
+    def trust_domain(self) -> TrustDomain:
+        """Returns the trust domain of the bundle. """
+        return self._trust_domain
+
+    def x509_authorities(self) -> Set[Certificate]:
+        """Returns the set of X.509 authorities in the bundle. """
+        return self._x509_authorities
+
+    def add_authority(self, x509_authority: Certificate) -> None:
+        """Adds an X.509 authority to the bundle. """
+        self._x509_authorities.add(x509_authority)
+
+    def remove_authority(self, x509_authority: Certificate) -> None:
+        """Removes an X.509 authority from the bundle. """
+        if not self._x509_authorities:
+            return
+        self._x509_authorities.remove(x509_authority)
+
+    @classmethod
+    def parse(cls, trust_domain: TrustDomain, bundle_bytes: bytes) -> 'X509Bundle':
+        """Parses an X.509 bundle from an array of bytes containing trusted authorities as PEM blocks.
+
+        Args:
+            trust_domain: A TrustDomain to associate to the bundle.
+            bundle_bytes: An array of bytes that represents a set of X.509 authorities.
+
+        Returns:
+            An instance of 'X509Bundle' with the X.509 authorities associated to the given trust domain.
+
+        Raises:
+            ParseBundleError: In case the set of x509_authorities cannot be parsed from the bundle_bytes.
+        """
+
+        if not trust_domain:
+            raise X509BundleError(EMPTY_DOMAIN_ERROR)
```

Comment:
nit: add X509BundleError to the docstring? A cpuple of other ocurrences of the same.

Reply:
Added the error to the docstrings and also moved the validation to the `__init__` method as it was repeated in multiple places and notably missing in the constructor.